File: pi-extension/src/oso_text.ts
/**
 * oso_text.ts — Helpers for assembling depth-limited text observation
 * bundles from OS Screen Observer responses.
 *
 * Port of /home/user/AutoGUI/oso_text.py.  Used by tools.ts for the
 * post-action injection on successful state-changing tool calls and by
 * emitRecoveryProbe() for the failure path.
 */

import type { ScreenObserverClient } from "./screen_observer_client.js";

interface TreeNode {
  name?: string;
  role?: string;
  bounds?: { x?: number; y?: number; width?: number; height?: number };
  children?: TreeNode[];
}

export interface TextBundle {
  description: string;
  sketch: string;
  treeText: string;
  depthUsed: number;
  truncated: boolean;
  scope: "active_window" | "screen";
}

export interface TextBundleOptions {
  windowIndex?: number;
  includeSketch: boolean;
  includeTree: boolean;
  treeStartDepth: number;
  treeMinDepth: number;
  treeMaxChars: number;
  maxChars: number;
}

/** Serialise an OSO tree as an indented role-name listing, depth-limited. */
export function flattenTree(node: TreeNode | undefined | null, depthLimit: number): string {
  if (!node) return "";
  const lines: string[] = [];
  const walk = (n: TreeNode, depth: number): void => {
    if (depth > depthLimit) return;
    const indent = "  ".repeat(depth);
    const role = (n.role ?? "").trim() || "?";
    const name = (n.name ?? "").trim();
    const namePart = name ? `  "${name}"` : "";
    const b = n.bounds;
    const boundsPart = b && b.width !== undefined
      ? `  [${Math.trunc(b.x ?? 0)},${Math.trunc(b.y ?? 0)} ${Math.trunc(b.width ?? 0)}x${Math.trunc(b.height ?? 0)}]`
      : "";
    lines.push(`${indent}${role}${namePart}${boundsPart}`);
    for (const child of n.children ?? []) walk(child, depth + 1);
  };
  walk(node, 0);
  return lines.join("\n");
}

/** Decrement depth until serialised tree fits under maxChars; hard-truncate
 *  at minDepth if it still overflows. */
export function trimTree(
  tree: TreeNode | undefined | null,
  opts: { startDepth: number; minDepth: number; maxChars: number },
): { text: string; depthUsed: number; truncated: boolean } {
  if (!tree) return { text: "", depthUsed: 0, truncated: false };
  let depth = Math.max(opts.minDepth, opts.startDepth);
  let text = "";
  while (depth >= opts.minDepth) {
    text = flattenTree(tree, depth);
    if (text.length <= opts.maxChars) {
      return { text, depthUsed: depth, truncated: false };
    }
    depth -= 1;
  }
  if (text.length > opts.maxChars) {
    text = `${text.slice(0, opts.maxChars)}\n… (truncated at ${opts.maxChars} chars)`;
    return { text, depthUsed: opts.minDepth, truncated: true };
  }
  return { text, depthUsed: opts.minDepth, truncated: false };
}

/** Fetch description/sketch/structure in parallel; assemble a length-bounded bundle. */
export async function buildOsoTextBundle(
  oso: ScreenObserverClient,
  opts: TextBundleOptions,
): Promise<TextBundle | null> {
  if (!oso || !oso.enabled) return null;
  const sketchP = opts.includeSketch ? oso.getSketch(opts.windowIndex) : Promise.resolve(null);
  const treeP = opts.includeTree ? oso.getStructure(opts.windowIndex) : Promise.resolve(null);
  const descP = oso.getDescription(opts.windowIndex);
  const [description, sketch, structure] = await Promise.all([descP, sketchP, treeP]);

  if (description === null && sketch === null && structure === null) return null;

  let descText = "";
  if (description) {
    descText = String(
      description.description ?? description.text ?? description.summary ?? "",
    );
  }
  let sketchText = "";
  if (sketch) {
    sketchText = String((sketch as { sketch?: string; text?: string }).sketch ?? (sketch as { text?: string }).text ?? "");
  }

  let treeText = "";
  let depthUsed = 0;
  let treeTruncated = false;
  if (structure) {
    const tree = (structure as { tree?: TreeNode }).tree ?? (structure as TreeNode);
    const trimmed = trimTree(tree, {
      startDepth: opts.treeStartDepth,
      minDepth: opts.treeMinDepth,
      maxChars: opts.treeMaxChars,
    });
    treeText = trimmed.text;
    depthUsed = trimmed.depthUsed;
    treeTruncated = trimmed.truncated;
  }

  const total = descText.length + sketchText.length + treeText.length;
  if (total > opts.maxChars) {
    let over = total - opts.maxChars;
    if (treeText && over > 0) {
      const cut = Math.min(treeText.length, over);
      treeText = treeText.slice(0, treeText.length - cut);
      if (cut > 0) {
        treeText += "\n… (truncated)";
        treeTruncated = true;
      }
      over -= cut;
    }
    if (over > 0 && sketchText) {
      const cut = Math.min(sketchText.length, over);
      sketchText = sketchText.slice(0, sketchText.length - cut);
      over -= cut;
    }
    if (over > 0 && descText) {
      descText = descText.slice(0, Math.max(0, descText.length - over));
    }
  }

  return {
    description: descText,
    sketch: sketchText,
    treeText,
    depthUsed,
    truncated: treeTruncated,
    scope: opts.windowIndex !== undefined ? "active_window" : "screen",
  };
}

File: oso_text.py
"""
oso_text.py — Shared helpers for building text observation bundles from
OS Screen Observer responses.

Used by backends/base.py (post-action injection) and agent.py (recovery
probe).  Keeping the depth-trimming + serialisation logic in one place
matches the same surface in pi-extension/src/oso_text.ts.
"""

from __future__ import annotations

from typing import Any


def flatten_tree(node: dict | None, depth_limit: int) -> str:
    """Serialise an OSO structure tree as an indented role-name listing.

    One element per line, truncated at depth_limit.  Compact format chosen
    over JSON to keep text-LLM context cheap and human-scannable.
    """
    if not node:
        return ""
    lines: list[str] = []

    def walk(n: dict, depth: int) -> None:
        if depth > depth_limit:
            return
        indent = "  " * depth
        role = (n.get("role") or "").strip() or "?"
        name = (n.get("name") or "").strip()
        bounds = n.get("bounds") or {}
        name_part = f'  "{name}"' if name else ""
        if bounds and bounds.get("width") is not None:
            bounds_part = (
                f"  [{int(bounds.get('x', 0))},{int(bounds.get('y', 0))} "
                f"{int(bounds.get('width', 0))}x{int(bounds.get('height', 0))}]"
            )
        else:
            bounds_part = ""
        lines.append(f"{indent}{role}{name_part}{bounds_part}")
        for child in n.get("children") or []:
            walk(child, depth + 1)

    walk(node, 0)
    return "\n".join(lines)


def trim_tree(
    tree: dict | None,
    *,
    start_depth: int,
    min_depth: int,
    max_chars: int,
) -> dict:
    """Serialise tree, lowering depth until output fits under max_chars.

    Returns {text, depth_used, truncated}.  Hard-truncates at min_depth
    when even that exceeds the budget.
    """
    if not tree:
        return {"text": "", "depth_used": 0, "truncated": False}
    depth = max(min_depth, start_depth)
    text = ""
    while depth >= min_depth:
        text = flatten_tree(tree, depth)
        if len(text) <= max_chars:
            return {"text": text, "depth_used": depth, "truncated": False}
        depth -= 1
    # min_depth still too big — hard-truncate.
    if len(text) > max_chars:
        text = text[:max_chars] + f"\n… (truncated at {max_chars} chars)"
        return {"text": text, "depth_used": min_depth, "truncated": True}
    return {"text": text, "depth_used": min_depth, "truncated": False}


async def build_text_bundle(
    oso: Any,
    *,
    window_index: int | None,
    include_sketch: bool,
    include_tree: bool,
    tree_start_depth: int,
    tree_min_depth: int,
    tree_max_chars: int,
    max_chars: int,
) -> dict | None:
    """Fetch description / sketch / structure in parallel and assemble a bundle.

    Returns None when OSO is unreachable; otherwise a dict with description,
    sketch, tree_text, depth_used, truncated, scope.  Total text length is
    capped to max_chars (description + sketch + tree combined).
    """
    if oso is None or not getattr(oso, "enabled", False):
        return None
    import asyncio

    sketch_task = (
        oso.get_sketch(window_index=window_index) if include_sketch else _none()
    )
    tree_task = (
        oso.get_structure(window_index=window_index) if include_tree else _none()
    )
    desc_task = oso.get_description(window_index=window_index)

    description, sketch, structure = await asyncio.gather(
        desc_task, sketch_task, tree_task
    )

    if description is None and sketch is None and structure is None:
        return None

    desc_text = ""
    if description:
        # OSO description payload varies — prefer a 'description' field, else
        # join known prose-like keys.  Keep it small.
        desc_text = (
            description.get("description")
            or description.get("text")
            or description.get("summary")
            or ""
        )
    sketch_text = ""
    if sketch:
        sketch_text = sketch.get("sketch") or sketch.get("text") or ""

    tree_text = ""
    depth_used = 0
    tree_truncated = False
    if structure:
        tree = structure.get("tree") or structure
        trimmed = trim_tree(
            tree,
            start_depth=tree_start_depth,
            min_depth=tree_min_depth,
            max_chars=tree_max_chars,
        )
        tree_text = trimmed["text"]
        depth_used = trimmed["depth_used"]
        tree_truncated = trimmed["truncated"]

    # Combined cap — drop tree first, then sketch, then description tail.
    total = len(desc_text) + len(sketch_text) + len(tree_text)
    if total > max_chars:
        over = total - max_chars
        if tree_text and over > 0:
            cut = min(len(tree_text), over)
            tree_text = tree_text[: len(tree_text) - cut]
            if cut > 0:
                tree_text += "\n… (truncated)"
                tree_truncated = True
            over -= cut
        if over > 0 and sketch_text:
            cut = min(len(sketch_text), over)
            sketch_text = sketch_text[: len(sketch_text) - cut]
            over -= cut
        if over > 0 and desc_text:
            desc_text = desc_text[: max(0, len(desc_text) - over)]

    scope = "active_window" if window_index is not None else "screen"
    return {
        "description": desc_text,
        "sketch": sketch_text,
        "tree_text": tree_text,
        "depth_used": depth_used,
        "truncated": tree_truncated,
        "scope": scope,
    }


async def _none():
    return None
